File: src/org/elixir_lang/utils/ElixirModulesUtil.java
package org.elixir_lang.utils;

import org.jetbrains.annotations.NotNull;

public class ElixirModulesUtil {
  @NotNull
  public static String elixirModuleNameToErlang(@NotNull String moduleName) {
    if (moduleName.equals("true") || moduleName.equals("false") || moduleName.equals("nil")) {
      return moduleName;
    } else if (moduleName.charAt(0) == ':') {
      return moduleName.substring(1);
    } else {
      return "Elixir." + moduleName;
    }
  }

  @NotNull
  public static String erlangModuleNameToElixir(@NotNull String moduleName) {
    if (moduleName.equals("true") || moduleName.equals("false") || moduleName.equals("nil")) {
      return moduleName;
    } else if (moduleName.startsWith("Elixir.")) {
      return moduleName.substring("Elixir.".length());
    } else {
      return ":" + moduleName;
    }
  }
}

File: src/org/elixir_lang/utils/ElixirModulesUtil.kt
package org.elixir_lang.utils

import org.elixir_lang.code.Identifier

object ElixirModulesUtil {
    // Matches a valid Elixir alias after stripping the "Elixir." prefix,
    // e.g. "Foo", "Foo.Bar", "Foo.Bar.Baz" — each segment starts with [A-Z]
    // and contains only [a-zA-Z0-9_].
    private val elixirAliasSegmentsRegex = Regex("([A-Z][a-zA-Z0-9_]*)(\\.[A-Z][a-zA-Z0-9_]*)*")

    fun elixirModuleNameToErlang(moduleName: String): String =
            when {
                moduleName == "true" || moduleName == "false" || moduleName == "nil" -> moduleName
                moduleName[0] == ':' -> moduleName.substring(1)
                else -> "Elixir." + moduleName
            }

    fun erlangModuleNameToElixir(moduleName: String): String =
            when {
                moduleName == "true" || moduleName == "false" || moduleName == "nil" -> moduleName
                moduleName.startsWith("Elixir.") -> {
                    val stripped = moduleName.removePrefix("Elixir.")
                    if (elixirAliasSegmentsRegex.matches(stripped)) {
                        stripped
                    } else {
                        // Not a valid alias (e.g. "Benchfella:tests") — render as a quoted atom
                        ":\"${stripped.replace("\\", "\\\\").replace("\"", "\\\"")}\""
                    }
                }
                else -> ":" + Identifier.inspectAsFunction(moduleName)
            }
}
